lib/irq.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/list.h>
+#include <metal/utilities.h>
+#include <metal/atomic.h>
+
+/** Soft IRQs mapping it stores which IRQ triggers the soft IRQ */
+struct metal_irq_map {
+	int source_irq_base; /**< Source IRQ which triggers the soft IRQ */
+	int target_irq_base; /**< target soft IRQ */
+	int irq_num; /**< number of IRQs of the mapping */
+};
+
+/** Indicate which IRQ has been allocated */
+#define METAL_IRQ_MAX 2048
+static unsigned long
+metal_irq_allocated[metal_div_round_up(METAL_IRQ_MAX, METAL_BITS_PER_ULONG)];
+
+/** List of registered IRQ controller */
+static METAL_DECLARE_LIST(irq_cntrs);
+
+#define METAL_IRQ_MAX_MAPPINGS 1
+static struct metal_irq_map irq_maps[METAL_IRQ_MAX_MAPPINGS];
+
+int metal_irq_allocate(int irq_base, int irq_num)
```

Comment:
This will not be needed when metal_irq_allocate() functionality is combined with metal_irq_register_controller()